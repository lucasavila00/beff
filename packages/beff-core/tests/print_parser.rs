#[cfg(test)]
mod tests {
    use std::{collections::BTreeSet, rc::Rc};

    use beff_core::{
        import_resolver::{parse_and_bind, FsModuleResolver},
        parser_extractor::BuiltDecoder,
        print::printer::ToWritableModules,
        schema_changes::print_ts_types,
        BeffUserSettings, BffFileName, EntryPoints, ExtractResult, FileManager, ParsedModule,
        Validator,
    };
    use swc_common::{Globals, GLOBALS};
    use swc_ecma_ast::TsType;
    struct TestFileManager {
        pub f: Rc<ParsedModule>,
    }

    impl FileManager for TestFileManager {
        fn get_or_fetch_file(&mut self, _name: &BffFileName) -> Option<Rc<ParsedModule>> {
            Some(self.f.clone())
        }

        fn get_existing_file(&self, _name: &BffFileName) -> Option<Rc<ParsedModule>> {
            Some(self.f.clone())
        }
    }

    struct TestResolver {}
    impl FsModuleResolver for TestResolver {
        fn resolve_import(&mut self, _module_specifier: &str) -> Option<BffFileName> {
            None
        }
    }
    fn parse_str(content: &str) -> Rc<ParsedModule> {
        let mut resolver = TestResolver {};
        let file_name = BffFileName::new("file.ts".into());
        GLOBALS.set(&Globals::new(), || {
            let res = parse_and_bind(&mut resolver, &file_name, content);
            res.expect("failed to parse")
        })
    }
    fn parse_api(it: &str) -> ExtractResult {
        let f = parse_str(it);
        let mut man = TestFileManager { f };
        let entry = EntryPoints {
            parser_entry_point: Some(BffFileName::new("file.ts".into())),
            schema_entry_point: None,
            settings: BeffUserSettings {
                custom_formats: BTreeSet::from_iter(vec!["password".to_string()]),
            },
        };
        beff_core::extract(&mut man, entry)
    }
    fn as_typescript_string_(validators: &[&Validator], built_decoders: &[BuiltDecoder]) -> String {
        let mut vs: Vec<(String, TsType)> = vec![];

        let mut sorted_validators = validators.iter().collect::<Vec<_>>();
        sorted_validators.sort_by(|a, b| a.name.cmp(&b.name));

        for v in sorted_validators {
            vs.push((v.name.clone(), v.schema.to_ts_type()));
        }

        let mut sorted_decoders = built_decoders.iter().collect::<Vec<_>>();
        sorted_decoders.sort_by(|a, b| a.exported_name.cmp(&b.exported_name));

        for v in sorted_decoders {
            vs.push((v.exported_name.clone(), v.schema.to_ts_type()));
        }

        print_ts_types(vs)
    }
    fn ok(from: &str) -> String {
        let p = parse_api(from);
        let errors = p.errors();

        if !errors.is_empty() {
            panic!("errors: {:?}", errors);
        }
        match p.parser {
            Some(v) => as_typescript_string_(
                &v.validators.iter().collect::<Vec<_>>(),
                v.built_decoders.as_ref().unwrap_or(&vec![]),
            ),
            None => panic!(),
        }
    }

    fn decoder(from: &str) -> String {
        let p = parse_api(from);
        let errors = p.errors();

        if !errors.is_empty() {
            panic!("errors: {:?}", errors);
        }
        match p.parser {
            Some(v) => {
                let res = ExtractResult {
                    parser: Some(v),
                    schema: None,
                };
                let m = res.to_module().expect("should be able to emit module");
                m.js_validators
            }
            None => panic!(),
        }
    }

    #[test]
    fn ok_either() {
        let from = r#"
    /**
     * @category model
     * @since 2.0.0
     */
    type Left<E1> = {
         _tag: 'Left',
         left: E1,
    }
    /**
     * @category model
     * @since 2.0.0
     */
    type Right<A1> = {
         _tag: 'Right',
         right: A1,
    }
    /**
     * @category model
     * @since 2.0.0
     */
    type Either<E, A> = Left<E> | Right<A>

    parse.buildParsers<{ A: Either<string, number> }>();

  "#;
        insta::assert_snapshot!(ok(from));
    }
    #[test]
    fn ok_omit() {
        let from = r#"

    type User = {
        name: string,
        age: number,
        email: string,
    }
    parse.buildParsers<{ A: Omit<User, 'age'> }>();

  "#;
        insta::assert_snapshot!(ok(from));
    }
    #[test]
    fn ok_pick() {
        let from = r#"

    type User = {
        name: string,
        age: number,
        email: string,
    }
    parse.buildParsers<{ A: Pick<User, 'age'> }>();

  "#;
        insta::assert_snapshot!(ok(from));
    }
    #[test]
    fn ok_omit2() {
        let from = r#"

    type User = {
        name: string,
        age: number,
        email: string,
    }
    parse.buildParsers<{ A: Omit<User, 'age'|'email'> }>();

  "#;
        insta::assert_snapshot!(ok(from));
    }
    #[test]
    fn ok_partial() {
        let from = r#"

    type User = {
        name: string,
        age: number,
        email: string,
    }
    parse.buildParsers<{ A: Partial<User> }>();

  "#;
        insta::assert_snapshot!(ok(from));
    }
    #[test]
    fn ok_partial2() {
        let from = r#"

    type User2 ={enabled: boolean} &Partial< {
        name: string,
        age: number,
        email: string,
    }>
    parse.buildParsers<{ A: User2 }>();

  "#;
        insta::assert_snapshot!(ok(from));
    }
    #[test]
    fn ok_partial3() {
        let from = r#"

    type User2 ={enabled: boolean} &Partial< {
        name: string,
        age: number,
        email: string,
    }>

    type ObjWithUsers = {enabled: boolean} & Partial<{a2: User2}>
    parse.buildParsers<{ A: ObjWithUsers }>();

  "#;
        insta::assert_snapshot!(ok(from));
    }
    #[test]
    fn ok_object() {
        insta::assert_snapshot!(ok(r#"
        parse.buildParsers<{ A: Object }>();
      "#));
    }
    #[test]
    fn ok_required() {
        insta::assert_snapshot!(ok(r#"

        type MaybeUser = {
            name?: string,
            age?: number,
        }
        parse.buildParsers<{ A: Required<MaybeUser> }>();
      "#));
    }
    #[test]
    fn ok_interface_extends() {
        insta::assert_snapshot!(ok(r#"

        interface User {
            name: string,
            age: number,
        }
        interface Admin extends User {
            role: string,
        }
        parse.buildParsers<{ Admin: Admin }>();
      "#));
    }
    #[test]
    fn ok_repro() {
        insta::assert_snapshot!(ok(r#"
        export type Settings = {
            a: string;
            level: "a" | "b";
            d: {
                tag: "d";
            };
        };
          
        export type SettingsUpdate = Settings["a" | "level" | "d"];
        parse.buildParsers<{ SettingsUpdate: SettingsUpdate }>();
      "#));
    }
    #[test]
    fn ok_mapped_type() {
        insta::assert_snapshot!(ok(r#"
        export type Mapped = {
            [K in "a" | "b"]: {
                value: K;
            };
        };
        parse.buildParsers<{ Mapped: Mapped }>();
      "#));
    }
    #[test]
    fn ok_mapped_type_optional() {
        insta::assert_snapshot!(ok(r#"
        export type Mapped = {
            [K in "a" | "b"]?: {
                value: K;
            };
        };
        parse.buildParsers<{ Mapped: Mapped }>();
      "#));
    }
    #[test]
    fn ok_mapped_type_repro() {
        insta::assert_snapshot!(ok(r#"
        type Obj = { a: string } & {d: string}
        type MappedKeys = keyof Obj;
        parse.buildParsers<{ MappedKeys: MappedKeys }>();
      "#));
    }
    #[test]
    fn ok_record_access() {
        insta::assert_snapshot!(ok(r#"
        export type Extra = Record<string, string>;
        type ExtraValue = Extra[string];

        parse.buildParsers<{ ExtraValue: ExtraValue }>();
      "#));
    }
    #[test]
    fn ok_record_union() {
        insta::assert_snapshot!(ok(r#"
        export type Extra = Record<'a'|'b', string>;

        parse.buildParsers<{ Extra: Extra }>();
      "#));
    }
    #[test]
    fn ok_array_spread() {
        insta::assert_snapshot!(ok(r#"
        const Arr1 = ["a", "b"] as const
        const Arr2 = [...Arr1, "c"] as const

        type Arr2C = typeof Arr2[number];

        parse.buildParsers<{ Arr2C: Arr2C }>();
      "#));
    }
    #[test]
    fn ok_array_spread_declare() {
        insta::assert_snapshot!(ok(r#"
        declare const Arr1 = ["a", "b"] as const
        declare const Arr2 = [...Arr1, "c"] as const

        type Arr2C = typeof Arr2[number];

        parse.buildParsers<{ Arr2C: Arr2C }>();
      "#));
    }
    #[test]
    fn ok_array_spread_declare2() {
        insta::assert_snapshot!(ok(r#"
        declare const AllArr1: ["a", "b"] 
        type Arr1 = typeof AllArr1[number];
        parse.buildParsers<{ Arr1: Arr1 }>();
      "#));
    }
    #[test]
    fn ok_array_spread_declare3() {
        insta::assert_snapshot!(ok(r#"
        export declare const AllArr1: ["a", "b"] 
        type Arr1 = typeof AllArr1[number];
        parse.buildParsers<{ Arr1: Arr1 }>();
      "#));
    }
    #[test]
    fn ok_array_spread2() {
        insta::assert_snapshot!(ok(r#"
        export const Arr1 = ["a", "b"] as const
        export type Arr1 = typeof Arr1[number]
        export const Arr2 = [...Arr1, "c"] as const

        type Arr2C = typeof Arr2[number];

        parse.buildParsers<{ Arr2C: Arr2C }>();
      "#));
    }
    #[test]
    fn ok_enum_member() {
        insta::assert_snapshot!(ok(r#"
        export enum Enum {
            A = "a",
            B = "b",
        }
        export type X = Enum.A

        parse.buildParsers<{ X: X }>();
      "#));
    }
    #[test]
    fn ok_enum_member2() {
        insta::assert_snapshot!(ok(r#"
        enum Enum {
            A = "a",
            B = "b",
        }
        type X = Enum.A

        parse.buildParsers<{ X: X }>();
      "#));
    }
    #[test]
    fn ok_discriminated_union() {
        insta::assert_snapshot!(decoder(
            r#"
        export type DiscriminatedUnion4 =
            | {
                type: "a";
                a: {
                  subType: "a1";
                  a1: string;
                };
              }
            | {
                type: "a";
                a: {
                  subType: "a2";
                  a2: string;
                };
              };
        parse.buildParsers<{ DiscriminatedUnion4: DiscriminatedUnion4 }>();
      "#
        ));
    }
    #[test]
    fn ok_exclude() {
        insta::assert_snapshot!(ok(r#"
        type A = "a" | "b";

        type B = "b" | "c";

        type X = Exclude<A, B>;

        parse.buildParsers<{ X: X }>();
      "#));
    }
    #[test]
    fn ok_exclude2() {
        insta::assert_snapshot!(ok(r#"
        type Shape =
        | { kind: "circle"; radius: number }
        | { kind: "square"; x: number }
        | { kind: "triangle"; x: number; y: number };
       
        type T3 = Exclude<Shape, { kind: "circle" }>
        parse.buildParsers<{ T3: T3 }>();
      "#));
    }
    #[test]
    fn ok_repro_3() {
        insta::assert_snapshot!(ok(r#"
      
        export interface IY {
            a: string
        }
        export interface IX {
            sizes?: IY;
        }
        // type IX2 = Required<IX>
        type T3 = IX[keyof IX]
        parse.buildParsers<{ T3: T3 }>();
      "#));
    }
    #[test]
    fn ok_recursive_tuple() {
        insta::assert_snapshot!(ok(r#"
        export type IX = [string, IX]
        type IX2 = IX[0]
        parse.buildParsers<{ IX2: IX2 }>();
      "#));
    }
    #[test]
    fn ok_conditional_type() {
        insta::assert_snapshot!(ok(r#"
        export type IX<T> = T extends true ? number : string
        type IX2 = IX<true>
        type IX3 = IX<false>
        parse.buildParsers<{ IX2: IX2, IX3: IX3 }>();
      "#));
    }
    #[test]
    fn ok_tpl_lit1() {
        insta::assert_snapshot!(ok(r#"
        export type IX = `a${string}b${number}c`
        parse.buildParsers<{ IX: IX }>();
      "#));
    }
    #[test]
    fn ok_tpl_lit2() {
        insta::assert_snapshot!(ok(r#"
        enum Abc {
            A = "A",
            B = "B",
        }
        export type IX = `${Abc}`
        parse.buildParsers<{ IX: IX }>();
      "#));
    }
    #[test]
    fn ok_repro4() {
        insta::assert_snapshot!(ok(r#"
        export const a = "a" as const;
        export const b = "b" as const;
        
        export const AllTs = [a, b] as const;
        export type AllTs = (typeof AllTs)[number];
        
        parse.buildParsers<{ AllTs: AllTs }>();
      "#));
    }
    #[test]
    fn ok_repro5() {
        insta::assert_snapshot!(ok(r#"
        export const ABC = {a: "b", c: "d"} as const satisfies Record<string, string>;
        export type AllTs = (keyof typeof ABC);

        parse.buildParsers<{ AllTs: AllTs }>();
      "#));
    }
    #[test]
    fn ok_void() {
        insta::assert_snapshot!(ok(r#"
        export type IX = void
        parse.buildParsers<{ IX: IX }>();
      "#));
    }
}
