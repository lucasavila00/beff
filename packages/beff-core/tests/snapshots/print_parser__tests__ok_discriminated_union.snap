---
source: packages/beff-core/tests/print_parser.rs
expression: "decoder(r#\"\n        export type DiscriminatedUnion4 =\n            | {\n                type: \"a\";\n                a: {\n                  subType: \"a1\";\n                  a1: string;\n                };\n              }\n            | {\n                type: \"a\";\n                a: {\n                  subType: \"a2\";\n                  a2: string;\n                };\n              };\n        parse.buildParsers<{ DiscriminatedUnion4: DiscriminatedUnion4 }>();\n      \"#)"
---
function ValidateDiscriminatedUnion4(ctx, input) {
    return (hoisted_DiscriminatedUnion4_12.validateAnyOfValidator.bind(hoisted_DiscriminatedUnion4_12))(ctx, input);
}
function ParseDiscriminatedUnion4(ctx, input) {
    return (hoisted_DiscriminatedUnion4_13.parseAnyOfParser.bind(hoisted_DiscriminatedUnion4_13))(ctx, input);
}
const validators = {
    DiscriminatedUnion4: ValidateDiscriminatedUnion4
};
const parsers = {
    DiscriminatedUnion4: ParseDiscriminatedUnion4
};
const hoisted_DiscriminatedUnion4_0 = new ConstDecoder("a1");
const hoisted_DiscriminatedUnion4_1 = new ObjectValidator({
    "a1": validateString,
    "subType": hoisted_DiscriminatedUnion4_0.validateConstDecoder.bind(hoisted_DiscriminatedUnion4_0)
}, null);
const hoisted_DiscriminatedUnion4_2 = new ObjectParser({
    "a1": parseIdentity,
    "subType": hoisted_DiscriminatedUnion4_0.parseConstDecoder.bind(hoisted_DiscriminatedUnion4_0)
}, null);
const hoisted_DiscriminatedUnion4_3 = new ConstDecoder("a");
const hoisted_DiscriminatedUnion4_4 = new ObjectValidator({
    "a": hoisted_DiscriminatedUnion4_1.validateObjectValidator.bind(hoisted_DiscriminatedUnion4_1),
    "type": hoisted_DiscriminatedUnion4_3.validateConstDecoder.bind(hoisted_DiscriminatedUnion4_3)
}, null);
const hoisted_DiscriminatedUnion4_5 = new ObjectParser({
    "a": hoisted_DiscriminatedUnion4_2.parseObjectParser.bind(hoisted_DiscriminatedUnion4_2),
    "type": hoisted_DiscriminatedUnion4_3.parseConstDecoder.bind(hoisted_DiscriminatedUnion4_3)
}, null);
const hoisted_DiscriminatedUnion4_6 = new ConstDecoder("a2");
const hoisted_DiscriminatedUnion4_7 = new ObjectValidator({
    "a2": validateString,
    "subType": hoisted_DiscriminatedUnion4_6.validateConstDecoder.bind(hoisted_DiscriminatedUnion4_6)
}, null);
const hoisted_DiscriminatedUnion4_8 = new ObjectParser({
    "a2": parseIdentity,
    "subType": hoisted_DiscriminatedUnion4_6.parseConstDecoder.bind(hoisted_DiscriminatedUnion4_6)
}, null);
const hoisted_DiscriminatedUnion4_9 = new ConstDecoder("a");
const hoisted_DiscriminatedUnion4_10 = new ObjectValidator({
    "a": hoisted_DiscriminatedUnion4_7.validateObjectValidator.bind(hoisted_DiscriminatedUnion4_7),
    "type": hoisted_DiscriminatedUnion4_9.validateConstDecoder.bind(hoisted_DiscriminatedUnion4_9)
}, null);
const hoisted_DiscriminatedUnion4_11 = new ObjectParser({
    "a": hoisted_DiscriminatedUnion4_8.parseObjectParser.bind(hoisted_DiscriminatedUnion4_8),
    "type": hoisted_DiscriminatedUnion4_9.parseConstDecoder.bind(hoisted_DiscriminatedUnion4_9)
}, null);
const hoisted_DiscriminatedUnion4_12 = new AnyOfValidator([
    hoisted_DiscriminatedUnion4_4.validateObjectValidator.bind(hoisted_DiscriminatedUnion4_4),
    hoisted_DiscriminatedUnion4_10.validateObjectValidator.bind(hoisted_DiscriminatedUnion4_10)
]);
const hoisted_DiscriminatedUnion4_13 = new AnyOfParser([
    hoisted_DiscriminatedUnion4_5.parseObjectParser.bind(hoisted_DiscriminatedUnion4_5),
    hoisted_DiscriminatedUnion4_11.parseObjectParser.bind(hoisted_DiscriminatedUnion4_11)
]);
